### ISSUE TITLE
opal/util/info.c can still truncate info keys

### ISSUE THREAD

**Issue by jsquyres** [open-mpi/ompi#5502]
On a build on master HEAD with gcc 7.3.0, I see the following (also seen by @rhc54 in https://www.mail-archive.com/devel@lists.open-mpi.org/msg20671.html):

```
91  In file included from info.c:46:0:
92  info.c: In function ‘opal_info_dup_mode’:
93  ../../opal/util/info.h:112:31: warning: ‘%s’ directive output may be truncated writing up to 36 bytes into a region of size 27 [-Wformat-truncation=]
94   #define OPAL_INFO_SAVE_PREFIX "_OMPI_IN_"
95                                 ^
96  info.c:212:22: note: in expansion of macro ‘OPAL_INFO_SAVE_PREFIX’
97                        OPAL_INFO_SAVE_PREFIX "%s", iterator->ie_key);
98                        ^~~~~~~~~~~~~~~~~~~~~
99  info.c:212:45: note: format string is defined here
100                       OPAL_INFO_SAVE_PREFIX "%s", iterator->ie_key);
101                                              ^~
102 info.c:211:18: note: ‘snprintf’ output between 10 and 46 bytes into a destination of size 36
103                   snprintf(savedkey, OPAL_MAX_INFO_KEY,
104                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
105                       OPAL_INFO_SAVE_PREFIX "%s", iterator->ie_key);
106                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I.e., it looks like #5204 didn't actually fix the problem.

@markalle Can you please fix?

I did not check to which back release branches this will also need to be applied.

**Comment by rhc54** [comment]
Also seeing these on my Mac using gcc-7.3:
```shell
In file included from /opt/local/lib/gcc7/gcc/x86_64-apple-darwin17/7.3.0/include-fixed/stdio.h:425:0,
                 from ../../opal/class/opal_list.h:71,
                 from ../../opal/util/info_subscriber.h:30,
                 from info_subscriber.c:45:
info_subscriber.c: In function 'opal_infosubscribe_change_info':
../../opal/util/info.h:112:31: warning: '%s' directive output may be truncated writing up to 36 bytes into a region of size 27 [-Wformat-truncation=]
 #define OPAL_INFO_SAVE_PREFIX "_OMPI_IN_"
                               ^
info_subscriber.c:264:13: note: in expansion of macro 'OPAL_INFO_SAVE_PREFIX'
             OPAL_INFO_SAVE_PREFIX "%s", key);
             ^~~~~~~~~~~~~~~~~~~~~
info_subscriber.c:264:36: note: format string is defined here
             OPAL_INFO_SAVE_PREFIX "%s", key);
                                    ^~
In file included from /opt/local/lib/gcc7/gcc/x86_64-apple-darwin17/7.3.0/include-fixed/stdio.h:425:0,
                 from ../../opal/class/opal_list.h:71,
                 from ../../opal/util/info_subscriber.h:30,
                 from info_subscriber.c:45:
info_subscriber.c:263:9: note: '__builtin_snprintf' output between 10 and 46 bytes into a destination of size 36
         snprintf(modkey, OPAL_MAX_INFO_KEY,
         ^
In file included from /opt/local/lib/gcc7/gcc/x86_64-apple-darwin17/7.3.0/include-fixed/stdio.h:425:0,
                 from ../../opal/class/opal_list.h:71,
                 from ../../opal/util/info.h:30,
                 from info.c:46:
info.c: In function 'opal_info_dup_mode.constprop':
../../opal/util/info.h:112:31: warning: '%s' directive output may be truncated writing up to 36 bytes into a region of size 27 [-Wformat-truncation=]
 #define OPAL_INFO_SAVE_PREFIX "_OMPI_IN_"
                               ^
info.c:212:22: note: in expansion of macro 'OPAL_INFO_SAVE_PREFIX'
                      OPAL_INFO_SAVE_PREFIX "%s", iterator->ie_key);
                      ^~~~~~~~~~~~~~~~~~~~~
info.c:212:45: note: format string is defined here
                      OPAL_INFO_SAVE_PREFIX "%s", iterator->ie_key);
                                             ^~
In file included from /opt/local/lib/gcc7/gcc/x86_64-apple-darwin17/7.3.0/include-fixed/stdio.h:425:0,
                 from ../../opal/class/opal_list.h:71,
                 from ../../opal/util/info.h:30,
                 from info.c:46:
info.c:211:18: note: '__builtin_snprintf' output between 10 and 46 bytes into a destination of size 36
                  snprintf(savedkey, OPAL_MAX_INFO_KEY,
```

**Comment by gpaulsen** [comment]
Reopen to PR to release branches.

**Comment by jsquyres** [comment]
Note that there are still other places long info keys can either overrun the array of produce amorphus runtime errors.

Coincidentally, I looked at this code in detail this past weekend.  I think that this code is due for a bit of revamping, for example:

1. A bunch of the comments do not match the code, which makes for very confusing reading.
1. Instead of having a magic prefix for "old" key values, just have a separate list.  It then avoids this whole "there's 2 different max key lengths" semantic mismatch.

My $0.02.

**Comment by gpaulsen** [comment]
I'll PR this to release branches.

**Comment by gpaulsen** [comment]
v4.0.x _WAS_ PRed here: https://github.com/open-mpi/ompi/pull/5757.
v3.1.x and v3.0.x has raised the bar as far as fixes going back there, and I doubt this fix raises to that level.  @jsquyres, @bwbarrett If you'd like me to PR back to v3.1.x this snprintf fix, please let me know and I can do so.
Assuming no, I'm closing this now.

**Comment by jsquyres** [comment]
At this point, I am fine with the fixes not coming back to v3.0.x and v3.1.x.
